llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp

```diff
@@ -1271,6 +1271,17 @@ Instruction *InstCombinerImpl::visitLShr(BinaryOperator &I) {
     return NewSub;
   }
 
+  // (sub nuw X, (Y << nuw Z)) >>u exact Z --> (X >>u exact Z) sub nuw Y
+  if (I.isExact() &&
+      match(Op0, m_OneUse(m_NUWSub(m_Value(X),
+                                   m_NUWShl(m_Value(Y), m_Specific(Op1)))))) {
+    Value *NewLshr = Builder.CreateLShr(X, Op1, "", /*isExact=*/true);
+    auto *NewSub = BinaryOperator::CreateNUWSub(NewLshr, Y);
+    NewSub->setHasNoSignedWrap(
```

Comment:
Please provide alive2 proof for nsw flag propagation.

Reply:
Done!

Reply:
Please double-check your proofs :)

Reply:
Done and done!

Reply:
```
define i16 @src1(i16 %x, i16 noundef %c, i16 %y) {
  %shl = shl nuw nsw i16 %x, %c
  %sub = sub nuw i16 %y, %shl
  %lshr = lshr exact i16 %sub, %c
  ret i16 %lshr
}

define i16 @tgt1(i16 %x, i16 noundef %c, i16 %y) {
  %tmp1 = lshr exact i16 %y, %c
  %sub = sub nuw i16 %tmp1, %x
  ret i16 %sub
}
```

`nsw` here is unnecessary.

Reply:
I fixed that.